File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    
    void dfs(TreeNode* root,TreeNode* parent,unordered_map<TreeNode*,TreeNode*>&gzip){
        
            if(root==NULL){
                return;
            }
            gzip[root]=parent;
            dfs(root->left,root,gzip);
            dfs(root->right,root,gzip);
    }
    
    
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        
            unordered_map<TreeNode*,TreeNode*>gzip;
            unordered_set<TreeNode*>s;
            dfs(root,NULL,gzip);
            queue<pair<TreeNode*,int>>q;
            q.push({target,0});
            int cnt=0;
            vector<int>ans;
            s.insert(target);
            while(!q.empty()){
                    auto temp=q.front();
                    q.pop();
                    if(temp.second==k){
                            ans.push_back(temp.first->val);        
                    }
                    if(temp.first->left and s.find(temp.first->left)==s.end()){
                            q.push({temp.first->left,temp.second+1});
                             s.insert(temp.first->left);
                    }
                    if(temp.first->right and s.find(temp.first->right)==s.end()){
                            q.push({temp.first->right,temp.second+1});
                            s.insert(temp.first->right);
                    }
                    if(gzip.find(temp.first)!=gzip.end() and gzip[temp.first]!=NULL and s.find(gzip[temp.first])==s.end()){
                            q.push({gzip[temp.first],temp.second+1});
                            s.insert(gzip[temp.first]);
                    }
                       
            }
        
             return ans; 
    }   
};

File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
    def __init__(self):
        
        self.graph={}
    
    
    def helper(self,root,parent):
        
        if root==None:
                return None
        
        self.graph[root.val]=[]
        if parent!=None:
                self.graph[root.val].append(parent.val)
        
        if root.left!=None:
                self.graph[root.val].append(root.left.val)
                self.helper(root.left,root)
                
        
        if root.right!=None:
                self.graph[root.val].append(root.right.val)
                self.helper(root.right,root)
                
    
    
    
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
    
            
            self.helper(root,None)
            queue=collections.deque([(target.val,0)])
            visited=[False]*(max(self.graph.keys())+1)
            #print(visited,max(self.graph.keys())+1)
            visited[target.val]=True
            
            ans=[]
            while(len(queue)):
                temp,dist = queue.popleft()
                if dist==k:
                        ans.append(temp)
                visited[temp]=True
                for item in self.graph[temp]:
                    if not visited[item]:
                            visited[item]=True
                            queue.append((item,dist+1))
            
                    
            return ans        
                    
                

                
### Time Complexity  
### Space Complexity